### DISCUSSION TITLE
In nextjs 13.4 how do we make POST calls from client component

### DISCUSSION THREAD

**Discussion by mighty98** [vercel/next.js#51139] [App Router]
Im using api routes where i am connecting to mongodb

I have a client component file which has a form to be submitted

I cannot avoid 'use client' as i useState
now build fails

`- wait compiling /api/subscribe/route (client and server)...
- error ./node_modules/saslprep/lib/memory-code-points.js:3:11
Module not found: Can't resolve 'fs'

https://nextjs.org/docs/messages/module-not-found

Import trace for requested module:
./node_modules/saslprep/index.js
./node_modules/mongodb/lib/deps.js
./node_modules/mongodb/lib/index.js
./lib/mongodb.ts
./app/api/subscribe/route.ts`

So how do we in NextJS13.4 make POST calls to add data to MongoDB atlas from a client component

**Comment by devjiwonchoi** [comment]
You can do this in various ways, but I would recommend you to try `Server Actions`. This allows the benefit of reducing dependency on the client-side Javascript.

> They enable server-side data mutations, reduced client-side JavaScript, and progressively enhanced forms.

Note that Server Actions is `alpha` stage, so you need to add a `experimental: serverActions` at your `next.config.js`.

```js
module.exports = {
  experimental: {
    serverActions: true,
  },
}
```

## Create an `actions.js` file inside your `/app` directory.

```js
// /app/actions

'use server'
 
export async function postToDB() {
  // ...
}
```

Note that you need a `'use server'` on the top of the file.

> If you're using a Server Action inside a Client Component, create your action in a separate file with the "use server" directive at the top of the file.

## Import it to your client-side `form`.

```js
"use client"
 
import { postToDB } from './actions';
 
export default function ClientComponent() {
  return (
    <form action={postToDB}>
      <button type="submit">Post to DB</button>
    </form>
  );
}
```

## Reference
- [Server Actions with Client Components](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions#with-client-components)

**Reply by mighty98** [reply]
Thank you so much for this.. 
By the way is there any way the client component can get the outcome of the server action. For instance if postToDB fails we would want to show an error message or take further actions

**Reply by devjiwonchoi** [reply]
You can add a custom `form-validation` and wrap your `action`.

```js
// lib/form-validation.ts

export function withValidate(action) {
  return (formData: FormData) => {
    'use server'
 
    const isValidData = verifyData(formData)
 
    if (!isValidData) {
      throw new Error('Error Message')
      // Or do something...
    }
 
    const data = process(formData)
    return action(data)
  }
}
```

```js
// /app/actions
'use server'
 
import { withValidate } from 'lib/form-validation'
 
export const action = withValidate((data) => {
  // ...
})
```

## Reference
- [Validation](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions#validation)

**Comment by icyJoseph** [comment]
Server Actions or just refactor the client component into its own thing, but I think you most likely want server actions.

**Reply by mighty98** [reply]
Thanks alot @icyJoseph

**Comment by mighty98** [comment]
Thank you so much for this.. 
By the way is there any way the client component can get the outcome of the server action. For instance if postToDB fails we would want to show an error message or take further actions
